pkg/scheduler/objects/application.go

```diff
@@ -266,12 +266,29 @@ func (sa *Application) timeoutPlaceholderProcessing() {
 	defer sa.Unlock()
 	switch {
 	// Case 1: if all app's placeholders are allocated, only part of them gets replaced, just delete the remaining placeholders
-	case (sa.IsRunning() || sa.IsStarting()) && !resources.IsZero(sa.allocatedPlaceholder):
+	case (sa.IsRunning() || sa.IsStarting() || sa.IsCompleting()) && !resources.IsZero(sa.allocatedPlaceholder):
```

Comment:
Why do we need here the `sa.IsCompleting()` check?
If the application is in completing state that means that the completing timer is started, so if it will timeout right after the placeholder timer, we might send the same allocation release request to the shim twice.
If we want to keep that check here, the state timer must be cleared.

Reply:
> If the application is in completing state that means that the completing timer is started, so if it will timeout right after the placeholder timer, we might send the same allocation release request to the shim twice.

There is no reason to assume that the timeouts will . The placeholder timer could time out 1 sec after entering the completing state.

What this check does is that it only removes placeholders that have not been replaced yet. If we move into the _Completing_ state we can do that with placeholder still allocated on the app. Any of those placeholders that have been marked as `released` have been communicated to the shim already for a replacement. They have a real allocation linked to it which is not confirmed yet as the core is waiting for the shim to confirm the release. If the replacement comes back from the shim we add the real allocation to the application. At that point we move back out of _Completing_ and we move back into the _Running_ state. As soon as that happens the timing out of the state _Completing_ will become a noop as the state is not the expected state.

That is the specific case we're trying to solve. The case 1 comment should be updated to cover all the use cases properly.

Reply:
Even if there is no reason to assume that the timeouts will follow quickly, it can easily happen. 

- Let's say we timeout the placeholders, and send the placeholder release request to the shim. As we know it can take a longer period for the shim to remove them, so in the core we will still have the placeholders marked as removed, until we have an answer from the shim.
- Right after that, the completing state times out(before we can get the replacement fro the shim). Since there we don't have this check for the already released placeholders, the core will send the remaining placeholders to the shim again.

So I think we need to mark the placeholders as released when we timeout the Completing state as well, and also we need to clear the completing state timer if we timeout the placeholders. There is no reason to keep it running until it will timeout.

Reply:
starting point of both cases:
- enter _Completing_ state, placeholder allocated
- placeholder timeout triggered: release all non replaced placeholder

case 1:
- in _Completing_, timeout placeholder triggered
- earlier replaced placeholder allocation is confirmed released by the shim, allocation is swapped
- application returns to _Running_
- --> placeholders that are timed out get removed
- --> placeholders that were replaced get swapped and app runs to a normal finish. All based on the shim responses

case 2:
- in _Completing_, timeout placeholder triggered
- no placeholder releases are confirmed
- _Completing_ timeout triggered, all placeholder will be released (again)
- wait for all placeholders to be released, all timers are cleaned up (existing behaviour)
- --> placeholders that are timed out get removed
- --> placeholders that were replaced get swapped and app runs to a normal finish. All based on the shim responses
- --> if all placeholders are removed and the app has not transitioned back to _Running_ it will transition to _Completed_

And yes in case it might happen that an original replacement request comes back before the second release requested. This is not a problem as that just simply triggers a change back into _Running_ for the app. All placeholders will be released either as a replacement or just a simple release. Changing the _Completing_ to _Completed_ transition itself is not what